
# 11ème école de bioinformatique AVIESAN - IFB - INSERM (EBAII 2022)

## Initiation au traitement des données de génomique obtenues par séquençage à haut débit

**Du 13-11-2022 au 18-11-2022 à la Station biologique de Roscoff**


## Web site

- Web pages: <https://ifb-elixirfr.github.io/EBAII/2022/ebaiin1>

From there, you can also find a link to download the whole repository with `git`.

****

## Supports de cours
### Linux

| Cours | Intervenants | Supports |
|----------------------------------|--------------|----------|
| Introduction à Linux + arborescence|  Thomas Denecker, Claire Toffano-Nioche | [gslides](https://docs.google.com/presentation/d/1181lS4gavBcVo0Yz70n6bawAvWdUlX-H3p0cFZ4Uadg/edit?usp=sharing) |
| Qualités et alignement | Thomas Denecker, Claire Toffano-Nioche | [gslides](https://docs.google.com/presentation/d/1ccuvsU7ibsn3GWMgp8wN0fOh7QmIgkIpCB-Mpt1vldU/edit?usp=sharing) |

### (Re)découverte de R

| Cours | Intervenants | Supports |
|-----------------------|--------------|-------------------------------------|
| Introduction à R | Thomas Denecker, Stevenn Volant | [[gslides](https://docs.google.com/presentation/d/1jK9-lZVmgSlPilTs4946_EiFaoQJmbvV/edit?usp=sharing&ouid=116446740554003676596&rtpof=true&sd=true)] |
| Données | | [[expression.txt](intro_R/expression.txt)] [[annotation.csv](intro_R/annotation.csv)] |
| Serveur RStudio IFB | | <https://rstudio.cluster.france-bioinformatique.fr/> |
| Serveur Jupyter IFB | | <https://jupyterhub.cluster.france-bioinformatique.fr/> |

### Exposés d'ouverture

| Cours | Intervenants | Supports |
|----------------------------------|--------------|----------|
| Long reads | Claude Thermes | [pdf](long_reads/20211123_ROSCOFF10_v14.pdf) |

### Visualisation

| Cours | Intervenants | Supports |
|----------------------------------|--------------|----------|
| Visualization of NGS Data using IGV (v2.4+) -- Documentation | Rachel Legendre, Pierre Pericard| [Intro](IGV/IGV.html) [Practical](IGV/IGV_practical.html) [data](IGV/data_igv.zip) |


### RNA-seq

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| Programme de l'atelier RNA-seq | |  [Program](RNAseq/README.md) |
| Bioinformatics | Emilie Drouineau | [pdf](RNA-seq/EBAI2020_RNAseq.pdf) [TP](RNA-seq/runme.sh) [TP corrigé](RNA-seq/runme_corrected.sh)|
| General statistics | Stevenn Volant | [pdf](RNAseq/RNASeq_differential_analysis.pdf) |
| Practice SARTools | Stevenn Volant | [ZIP archive](RNAseq/Practice_SARTools.zip) |
| Gene set analysis | Emilie Drouineau | [pdf](RNAseq/GSEA_Theory.pdf) [R script](RNAseq/GSEA_TP.R)|
| De novo assembly | Erwan Corre | [pdf](RNAseq/RNASeq_denovo_EBAII2021_red.pdf)  |
| Workflow | Pierre Pericard | [pdf](RNAseq/RNA-Workflow_Conclusion.pdf)  |

### Variants

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| Programme de l'atelier Variant | | [Program](DNA-seq/) |
| Introduction & Processing Post-Alignement | Nadia Bessoltane & Vivien Deshaies |[pdf](DNA-seq/1-Intro_Processing-Post-Alignement.pdf) [gslides](https://docs.google.com/presentation/d/1Z9HlepuI017lGUkFlpwLxdvx4QARG1RQ/) |
| Variant calling & Annotation | Vivien Deshaies | [pdf](DNA-seq/2-Variant-calling_Annotation.pdf) [gslides](https://docs.google.com/presentation/d/1JJffLPchxxhy1SFIi-0UedWvn5-n9VgK/) |
| Variants structuraux | Vivien Deshaies & Nadia Bessoltane | [pdf](DNA-seq/3-Variants-Structuraux.pdf) [gslides](https://docs.google.com/presentation/d/1tJn2-RMWkn7nLrcOubJPvMKx9A8WT2KP/) |
| Applications sous R | Nadia Bessoltane | [pdf](DNA-seq/?) [gslides](https://docs.google.com/presentation/d/1oOiJfcdnbIaY3G3F_CI-1nt-Op0W0aR8/) |
| Manipuler des VCF sous R | Nadia Bessoltane | [pdf](DNA-seq/?) [gslides](https://docs.google.com/presentation/d/10W-EipRAc6146UZwVIRQCvS3Riau3QO4/) [html](DNA-seq/?) |
| Workflow et conclusion | Nadia Bessoltane & Vivien Deshaies | [pdf](DNA-seq/6-Workflow_Conclusion.pdf) [gslides](https://docs.google.com/presentation/d/1-nON-453OfocMrXJ95fMc-JaU9KRcNnB/) |


### ChIP-seq

| Support | Lien |
|----------------------------------|------------------|
| Partie théorique | [gslides](https://docs.google.com/presentation/d/1v1_yncOrakL0vH_GvykNtuChM6TXI0V9c7pruSe9QOA/edit?usp=sharing) |
| Partie pratique | [TD](chip-seq/hands-on/) |

| Cours | Intervenants |
|----------------------------------|------------------|
| Programme de l'atelier ChIP-seq | [Program](chip-seq/) |
| Design expérimental, Contrôle qualité | Stéphanie Le Gras |
| Alignement | Stéphanie Le Gras |
| Contrôle qualité des alignements, visualisation | Tao Ye |
| Peak-calling | Tao Ye |
| Analyse des motifs | Stéphanie Le Gras |
| Annotation des pics | Tao Ye |


### Croisement et intégration des données

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| Croisement de données | Claire Toffano-Nioche, Pauline François | [gslides](https://docs.google.com/presentation/d/1IE8uBw0NPEliLF79s7hV1T2LaRLxoYzBttWn6B_4pPI/edit?usp=sharing) [soluce](https://docs.google.com/document/d/1oHT_OfGeAMB-61igEoXfC0JhqJMD0dowlvNX6cv8C1Y/edit?usp=sharing) [soluce_v2](croisement_donnees/soluce_long.sh) |

### Initiation Slurm

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| IFB Core Cluster - Introduction slides | Gildas Le Corguillé, Julien Seiler |[html](https://github.com/IFB-ElixirFr/EBAII/blob/master/2022/ebaiin1/cluster_slurm/20211125%20-%20EBAII2021-%20Slides%20cluster.pdf) |
