
# 9ème école de bioinformatique AVIESAN - IFB - INSERM (EBAII 2020)

## Initiation au traitement des données de génomique obtenues par séquençage à haut débit

**Du 04-10-2020 au 09-10-2020 à la Station biologique de Roscoff**


## Web site

- Web pages: <https://ifb-elixirfr.github.io/EBAII/2020/>

From there, you can also find a link to download the whole repository with `git`.

****

## Supports de cours

### Linux

| Cours | Intervenants | Supports |
|----------------------------------|--------------|----------|
| Introduction à Linux + arborescence| Denis Puthier, Gildas Le Corguillé, Claire Toffano-Nioche |  [pdf](intro_linux/intro_linux_ebaii_2020.pdf) [gslides](https://docs.google.com/presentation/d/1acl8ekE_FwzWbYlct8dEG-IeFV0d99UaHUJJWse0S5s/edit?usp=sharing) |
| Qualités des lectures | Denis Puthier, Gildas Le Corguillé, Claire Toffano-Nioche | [pdf](preprocessing_mapping/preprocessing_mapping_ebaii_2020.pdf) [gslides](https://docs.google.com/presentation/d/1YSa5WAL2g85e4R0qDvc2g1w4MFZfbVj82ZaS-X_2Ivs/edit?usp=sharing) |
| Read mapping | Denis Puthier, Gildas Le Corguillé, Claire Toffano-Nioche | [pdf](preprocessing_mapping/preprocessing_mapping_ebaii_2020.pdf) [gslides](https://docs.google.com/presentation/d/1YSa5WAL2g85e4R0qDvc2g1w4MFZfbVj82ZaS-X_2Ivs/edit?usp=sharing) |

### R

| Cours | Intervenants | Supports |
|-----------------------|--------------|-------------------------------------|
| Introduction à R | Thomas Denecker, Jacques van Helden, Hugo Varet | [[pdf](R/Intro_R_2020.pdf)] [[gslides](https://docs.google.com/presentation/d/1wJHUGDIip5EE8NL7mLSZFfYR9c-u2AAKKh30SQjuL-Q/edit?usp=sharing)] |
| Script  | | [[R script](R/script_intro_R_2020.r)] |
| Données | | [[expression.txt](R/expression.txt)] [[annotation.csv](R/annotation.csv)] |
| Serveur RStudio IFB | | <https://rstudio.cluster.france-bioinformatique.fr/> |
| Serveur Jupyter IFB | | <https://jupyterhub.cluster.france-bioinformatique.fr/> |

### Exposés d'ouverture

| Cours | Intervenants | Supports |
|----------------------------------|--------------|----------|
| Single-cell | Bastien Job | [pdf](single_cell/Intro_Single-Cell_20201005.pdf) [gslides](https://drive.google.com/file/d/1eFYJ-kRoUnYg_iCZG5pLd6zFDT6br0V8/view?usp=sharing) |
| Long reads | Claude Thermes | [ppt](long_reads/20201006_THERMES_ROSCOFF9_V3.pptx) |

### Visualisation

| Cours | Intervenants | Supports |
|----------------------------------|--------------|----------|
| Visualization of NGS Data using IGV (v2.4+) -- Documentation | Elodie Girard , Rachel Legendre, Olivier Quenez, Bastien Job| [Intro](IGV/IGV.html) [Practical](IGV/IGV_practical.html) [data](IGV/data_igv.zip) |


### RNA-seq

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| Programme de l'atelier RNA-seq | |  [Program](RNA-seq/) |
| Bioinformatics | Rachel Legendre | [pdf](RNA-seq/EBAI2020_RNAseq.pdf) [TP](RNA-seq/runme.sh) [TP corrigé](RNA-seq/runme_corrected.sh)|
| General statistics | Hugo Varet | [pdf](RNA-seq/RNASeq_differential_analysis.pdf) |
| Practice SARTools | Hugo Varet | [ZIP archive](RNA-seq/TP_differential_analysis/Practice_SARTools.zip) |
| Gene set analysis | Thibault Dayris | [pdf](RNA-seq/GSEA/GSEA.pdf) [R script](RNA-seq/GSEA/GSEA.R)|
| De novo assembly | Erwan Corre | [pdf](RNA-seq/RNASeq_denovo_EBAII2020_red.pdf)  |

### Variants

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| Programme de l'atelier Variant | | [Program](DNA-seq/) |
| Introduction  | Maria Bernard, Olivier Rué |[pdf](DNA-seq/1-Introduction-Atelier-Variant.pdf) [gslides](https://docs.google.com/presentation/d/1xOwOgrEDZvF-XV9waKNIpTXbGNLDCb5_APL5gXFSycc/edit?usp=sharing)   |
|  Processing Post-Alignement | Olivier Rué | [pdf](DNA-seq/2-Processing-Post-Alignement.pdf) [gslides](https://docs.google.com/presentation/d/1cfdGX-3rjhNAFGSBsk2NhBU44f7A4lkFeS_8ptdlLEw/edit?usp=sharing) |
| Variant calling | Nadia Bessoltane | [pdf](DNA-seq/3-Variant-calling.pdf) [gslides](https://docs.google.com/presentation/d/1GbfYjfXeeow6wkRRLR0MmDie5OTd1j3Cg02xgfGlUeI/edit?usp=sharing) |
| Filtrage et annotation | Olivier Rué | [pdf](DNA-seq/4-Filtrage-Annotation.pdf) [gslides](https://docs.google.com/presentation/d/17INp7S2o4IDxWOklkThMlut8ij1t4CZWGB8v73vmQc0/edit?usp=sharing) |
| Variants structuraux | Olivier Quenez | [pdf](DNA-seq/5-Variants-Structuraux.pdf) [gslides](https://docs.google.com/presentation/d/17CfyWqUNaCL1LnPgwX8J8srOKzZHlqxQj_e1j0Xy8-w/edit?usp=sharing) |
| Applications sous R | Maria Bernard |[html](DNA-seq/EBAII2020_variants.html)|
| Workflow et conclusion | Maria Bernard | [pdf](DNA-seq/6-Workflow_Conclusion.pdf) [gslides](https://docs.google.com/presentation/d/1nFRswgW8G1IAB-JIqp6D4tX_b94lKCWQ8kre6AZfClo/edit?usp=sharing) |
| Schéma récapitulatif | | [Schema](DNA-seq/variantCalling_pipeline.pdf) |

### ChIP-seq

| Support | Lien |
|----------------------------------|------------------|
| Partie théorique | [gslides](https://docs.google.com/presentation/d/1v1_yncOrakL0vH_GvykNtuChM6TXI0V9c7pruSe9QOA/edit?usp=sharing) |
| Partie pratique | [TD](chip-seq/hands-on/) |

| Cours | Intervenants |
|----------------------------------|------------------|
| Programme de l'atelier ChIP-seq | [Program](chip-seq/) |
| Design expérimental, Contrôle qualité | Stéphanie Le Gras |
| Alignement | Stéphanie Le Gras |
| Contrôle qualité des alignements, visualisation | Tao Ye |
| Peak-calling | Stéphanie Le Gras |
| Analyse des motifs | Elodie Darbo |
| Annotation des pics | Tao Ye |


### Croisement et intégration des données

| Cours | Intervenants | Supports |
|----------------------------------|------------------|----------|
| Croisement de données | Bastien Job, Claire Toffano-Nioche, Matthias Zytnicki | [gslides](https://docs.google.com/presentation/d/15qE95kW9H6fAha-2LnCVpMZxEwUDKZyedESeopzR704/edit?usp=sharing) [pdf](croisement_donnees/Cours.pdf) |
| Intégration de données | Laura Cantini |[pdf](croisement_donnees/CANTINI_NGS_AVIESAN_IFB.pdf) |

*****
