{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4e79c125-4298-44b7-b7cc-5e1d2f591912",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Demander de l'aide\n",
    "\n",
    "###### Equipe pédagogique : P. François, M. Genete, C. Toffano-Nioche, E. Jacquemet, G. Le Corguillé, J. Seiler, D. Puthier, T. Denecker et tout le staff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "392ece94-a8d9-498f-9986-4aedc439f7e8",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "source": [
    "Que nous débutions ou que nous soyons des bioinformaticiens chevronnés, il y a toujours un moment où nous sommes bloqués."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246787e0-ce12-4412-a146-e1ed9bc7c75b",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Quelles sont les options et arguments de mon outils ? Comment lui indiquer ma donnée d'entrée ? Comment extraire l'information X de mon fichier ? Il existe plusieurs moyens d'obtenir des réponses à (presque) toutes vos questions !"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3318ae97-9d1b-4111-994a-f844d2759aa3",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "## Pour les commandes bash-iques \n",
    "###### (Oui, on fait de l'informatique donc on continue les jeux de mots de qualité)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "605981e2-5740-43b3-8a8e-34ce9a052bb4",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "source": [
    "<code>man</code> est la commande qu'il vous faut. Elle veut litteralement dire **man**uel donc vous permettra d'afficher toute l'aide dont vous avez besoin pour la grande majorité des commandes inclus par défaut dans le langage _bash_.\n",
    "\n",
    "Pour l'utiliser : <code>man commande_mystere</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57ebb8e6-3a52-44a5-9c6a-05918a07b257",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Essayons. <code>whoami</code> est une commande possédant l'un des plus petits manuels. Mais que fait-elle ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e150c49b-a57f-4467-b430-3d54021e4a12",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    },
    "tags": []
   },
   "source": [
    "Dans le cadre ci-dessous, tapez la commande pour afficher l'aide de \"whoami\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4f36a4b0-73a9-4a56-8303-f270fc080021",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WHOAMI(1)                        User Commands                       WHOAMI(1)\n",
      "\n",
      "NAME\n",
      "       whoami - print effective userid\n",
      "\n",
      "SYNOPSIS\n",
      "       whoami [OPTION]...\n",
      "\n",
      "DESCRIPTION\n",
      "       Print  the  user  name  associated  with the current effective user ID.\n",
      "       Same as id -un.\n",
      "\n",
      "       --help display this help and exit\n",
      "\n",
      "       --version\n",
      "              output version information and exit\n",
      "\n",
      "AUTHOR\n",
      "       Written by Richard Mlynarik.\n",
      "\n",
      "REPORTING BUGS\n",
      "       GNU coreutils online help: <https://www.gnu.org/software/coreutils/>\n",
      "       Report   whoami   translation    bugs    to    <https://translationpro‐\n",
      "       ject.org/team/>\n",
      "\n",
      "COPYRIGHT\n",
      "       Copyright  ©  2018  Free Software Foundation, Inc.  License GPLv3+: GNU\n",
      "       GPL version 3 or later <https://gnu.org/licenses/gpl.html>.\n",
      "       This is free software: you are free  to  change  and  redistribute  it.\n",
      "       There is NO WARRANTY, to the extent permitted by law.\n",
      "\n",
      "SEE ALSO\n",
      "       Full documentation at: <https://www.gnu.org/software/coreutils/whoami>\n",
      "       or available locally via: info '(coreutils) whoami invocation'\n",
      "\n",
      "GNU coreutils 8.30              September 2019                       WHOAMI(1)\n"
     ]
    }
   ],
   "source": [
    "# Afficher l'aide de whoami\n",
    "man whoami"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d38420a3-cb08-40b8-a491-36ddd504ef37",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "source": [
    "On retrouve donc d'abord le nom de la commande et un résumé de ce qu'elle fait.\n",
    "\n",
    "Puis le \"synopsis\" vous expliquant comment l'écrire. \n",
    "\n",
    "La description vous liste l'ensemble des options existantes, leur but et leur utilisation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83a7e9eb-fe51-41fb-8797-5bb470f75c08",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Ici la commande whoami permet donc d'afficher votre nom d'utilisateur courant sans besoin de mettre d'argument.\n",
    "\n",
    "Vous pouvez essayer cette commande et vérifier que <code>man</code> ne vous a pas menti :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3e25ecf9-d230-4d5b-a119-533d1439c706",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pfrancois\n"
     ]
    }
   ],
   "source": [
    "# Utilisation de la commande whoami\n",
    "whoami"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05bfe648-93f2-4465-999c-be29ba3bbf75",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "## Pour un outil autre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b004233-1b98-445c-a2b2-19197d6e4068",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "source": [
    "Vous avez trouvé un outil qui, théoriquement, fait ce dont vous avez besoin. Vous voulez en savoir plus !\n",
    "\n",
    "Si nous affirmons que cette option est à 100% standardisée, nous trouverons un contre-exemple. Disons donc que vous ne risquez pas de trop vous tromper en utilisant l'option <code>--help</code> (ou sa forme courte <code>-h</code>)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dabe2aa-949c-4630-8576-51ea587ce73d",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Testons avec le fameux \"fastqc\" vu précédemment afin d'enfin lever le voile sur ce que fait cet outil."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f8e73e1-023f-4ead-8169-87f204a91d42",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Faites un <code>module load</code> de <code>fastqc/0.11.9</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "778c58fe-5740-482e-9eaa-947ffa81cab7",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import de l'outil fastqc en version 0.11.9\n",
    "module load fastqc/0.11.9"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da92c9c-1a89-4bfa-b969-fd48a12933bd",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Affichez son aide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3b8b2cb4-58ce-4c13-8fbb-87e931518170",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "            FastQC - A high throughput sequence QC analysis tool\n",
      "\n",
      "SYNOPSIS\n",
      "\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\n",
      "\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \n",
      "           [-c contaminant file] seqfile1 .. seqfileN\n",
      "\n",
      "DESCRIPTION\n",
      "\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\n",
      "    control report consisting of a number of different modules, each one of \n",
      "    which will help to identify a different potential type of problem in your\n",
      "    data.\n",
      "    \n",
      "    If no files to process are specified on the command line then the program\n",
      "    will start as an interactive graphical application.  If files are provided\n",
      "    on the command line then the program will run with no user interaction\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\n",
      "    analysis pipeline.\n",
      "    \n",
      "    The options for the program as as follows:\n",
      "    \n",
      "    -h --help       Print this help file and exit\n",
      "    \n",
      "    -v --version    Print the version of the program and exit\n",
      "    \n",
      "    -o --outdir     Create all output files in the specified output directory.\n",
      "                    Please note that this directory must exist as the program\n",
      "                    will not create it.  If this option is not set then the \n",
      "                    output file for each sequence file is created in the same\n",
      "                    directory as the sequence file which was processed.\n",
      "                    \n",
      "    --casava        Files come from raw casava output. Files in the same sample\n",
      "                    group (differing only by the group number) will be analysed\n",
      "                    as a set rather than individually. Sequences with the filter\n",
      "                    flag set in the header will be excluded from the analysis.\n",
      "                    Files must have the same names given to them by casava\n",
      "                    (including being gzipped and ending with .gz) otherwise they\n",
      "                    won't be grouped together correctly.\n",
      "                    \n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\n",
      "                    this mode you can pass in directories to process and the program\n",
      "                    will take in all fast5 files within those directories and produce\n",
      "                    a single output file from the sequences found in all files.                    \n",
      "                    \n",
      "    --nofilter      If running with --casava then don't remove read flagged by\n",
      "                    casava as poor quality when performing the QC analysis.\n",
      "                   \n",
      "    --extract       If set then the zipped output file will be uncompressed in\n",
      "                    the same directory after it has been created.  By default\n",
      "                    this option will be set if fastqc is run in non-interactive\n",
      "                    mode.\n",
      "                    \n",
      "    -j --java       Provides the full path to the java binary you want to use to\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\n",
      "                    your path.\n",
      "                   \n",
      "    --noextract     Do not uncompress the output file after creating it.  You\n",
      "                    should set this option if you do not wish to uncompress\n",
      "                    the output when running in non-interactive mode.\n",
      "                    \n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\n",
      "                    show data for every base in the read.  WARNING: Using this\n",
      "                    option will cause fastqc to crash and burn if you use it on\n",
      "                    really long reads, and your plots may end up a ridiculous size.\n",
      "                    You have been warned!\n",
      "                    \n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\n",
      "                    to be shown in the report.  As long as you set this to a value\n",
      "                    greater or equal to your longest read length then this will be\n",
      "                    the sequence length used to create your read groups.  This can\n",
      "                    be useful for making directly comaparable statistics from \n",
      "                    datasets with somewhat variable read lengths.\n",
      "                    \n",
      "    -f --format     Bypasses the normal sequence file format detection and\n",
      "                    forces the program to use the specified format.  Valid\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\n",
      "                    \n",
      "    -t --threads    Specifies the number of files which can be processed\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\n",
      "                    memory so you shouldn't run more threads than your\n",
      "                    available memory will cope with, and not more than\n",
      "                    6 threads on a 32 bit machine\n",
      "                  \n",
      "    -c              Specifies a non-default file which contains the list of\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\n",
      "                    The file must contain sets of named contaminants in the\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\n",
      "                    be ignored.\n",
      "\n",
      "    -a              Specifies a non-default file which contains the list of\n",
      "    --adapters      adapter sequences which will be explicity searched against\n",
      "                    the library. The file must contain sets of named adapters\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\n",
      "                    will be ignored.\n",
      "                    \n",
      "    -l              Specifies a non-default file which contains a set of criteria\n",
      "    --limits        which will be used to determine the warn/error limits for the\n",
      "                    various modules.  This file can also be used to selectively \n",
      "                    remove some modules from the output all together.  The format\n",
      "                    needs to mirror the default limits.txt file found in the\n",
      "                    Configuration folder.\n",
      "                    \n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\n",
      "                    length is 7 if not specified.\n",
      "                    \n",
      "   -q --quiet       Supress all progress messages on stdout and only report errors.\n",
      "   \n",
      "   -d --dir         Selects a directory to be used for temporary files written when\n",
      "                    generating report images. Defaults to system temp directory if\n",
      "                    not specified.\n",
      "                    \n",
      "BUGS\n",
      "\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\n",
      "                   \n",
      "    \n"
     ]
    }
   ],
   "source": [
    "# Version courte\n",
    "fastqc -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "98967b41-93d1-4135-90dd-4f6b0a56b2fb",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "            FastQC - A high throughput sequence QC analysis tool\n",
      "\n",
      "SYNOPSIS\n",
      "\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\n",
      "\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \n",
      "           [-c contaminant file] seqfile1 .. seqfileN\n",
      "\n",
      "DESCRIPTION\n",
      "\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\n",
      "    control report consisting of a number of different modules, each one of \n",
      "    which will help to identify a different potential type of problem in your\n",
      "    data.\n",
      "    \n",
      "    If no files to process are specified on the command line then the program\n",
      "    will start as an interactive graphical application.  If files are provided\n",
      "    on the command line then the program will run with no user interaction\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\n",
      "    analysis pipeline.\n",
      "    \n",
      "    The options for the program as as follows:\n",
      "    \n",
      "    -h --help       Print this help file and exit\n",
      "    \n",
      "    -v --version    Print the version of the program and exit\n",
      "    \n",
      "    -o --outdir     Create all output files in the specified output directory.\n",
      "                    Please note that this directory must exist as the program\n",
      "                    will not create it.  If this option is not set then the \n",
      "                    output file for each sequence file is created in the same\n",
      "                    directory as the sequence file which was processed.\n",
      "                    \n",
      "    --casava        Files come from raw casava output. Files in the same sample\n",
      "                    group (differing only by the group number) will be analysed\n",
      "                    as a set rather than individually. Sequences with the filter\n",
      "                    flag set in the header will be excluded from the analysis.\n",
      "                    Files must have the same names given to them by casava\n",
      "                    (including being gzipped and ending with .gz) otherwise they\n",
      "                    won't be grouped together correctly.\n",
      "                    \n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\n",
      "                    this mode you can pass in directories to process and the program\n",
      "                    will take in all fast5 files within those directories and produce\n",
      "                    a single output file from the sequences found in all files.                    \n",
      "                    \n",
      "    --nofilter      If running with --casava then don't remove read flagged by\n",
      "                    casava as poor quality when performing the QC analysis.\n",
      "                   \n",
      "    --extract       If set then the zipped output file will be uncompressed in\n",
      "                    the same directory after it has been created.  By default\n",
      "                    this option will be set if fastqc is run in non-interactive\n",
      "                    mode.\n",
      "                    \n",
      "    -j --java       Provides the full path to the java binary you want to use to\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\n",
      "                    your path.\n",
      "                   \n",
      "    --noextract     Do not uncompress the output file after creating it.  You\n",
      "                    should set this option if you do not wish to uncompress\n",
      "                    the output when running in non-interactive mode.\n",
      "                    \n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\n",
      "                    show data for every base in the read.  WARNING: Using this\n",
      "                    option will cause fastqc to crash and burn if you use it on\n",
      "                    really long reads, and your plots may end up a ridiculous size.\n",
      "                    You have been warned!\n",
      "                    \n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\n",
      "                    to be shown in the report.  As long as you set this to a value\n",
      "                    greater or equal to your longest read length then this will be\n",
      "                    the sequence length used to create your read groups.  This can\n",
      "                    be useful for making directly comaparable statistics from \n",
      "                    datasets with somewhat variable read lengths.\n",
      "                    \n",
      "    -f --format     Bypasses the normal sequence file format detection and\n",
      "                    forces the program to use the specified format.  Valid\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\n",
      "                    \n",
      "    -t --threads    Specifies the number of files which can be processed\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\n",
      "                    memory so you shouldn't run more threads than your\n",
      "                    available memory will cope with, and not more than\n",
      "                    6 threads on a 32 bit machine\n",
      "                  \n",
      "    -c              Specifies a non-default file which contains the list of\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\n",
      "                    The file must contain sets of named contaminants in the\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\n",
      "                    be ignored.\n",
      "\n",
      "    -a              Specifies a non-default file which contains the list of\n",
      "    --adapters      adapter sequences which will be explicity searched against\n",
      "                    the library. The file must contain sets of named adapters\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\n",
      "                    will be ignored.\n",
      "                    \n",
      "    -l              Specifies a non-default file which contains a set of criteria\n",
      "    --limits        which will be used to determine the warn/error limits for the\n",
      "                    various modules.  This file can also be used to selectively \n",
      "                    remove some modules from the output all together.  The format\n",
      "                    needs to mirror the default limits.txt file found in the\n",
      "                    Configuration folder.\n",
      "                    \n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\n",
      "                    length is 7 if not specified.\n",
      "                    \n",
      "   -q --quiet       Supress all progress messages on stdout and only report errors.\n",
      "   \n",
      "   -d --dir         Selects a directory to be used for temporary files written when\n",
      "                    generating report images. Defaults to system temp directory if\n",
      "                    not specified.\n",
      "                    \n",
      "BUGS\n",
      "\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\n",
      "                   \n",
      "    \n"
     ]
    }
   ],
   "source": [
    "# Version longue\n",
    "fastqc --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f292fafd-b6ca-483b-a115-9815efb13498",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "## Quand rien n'va plus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c09d25a-e465-4289-9ba2-a80e2a1ac5fd",
   "metadata": {
    "slideshow": {
     "slide_type": "subslide"
    },
    "tags": []
   },
   "source": [
    "Cette solution peut paraître simpliste et évidente mais elle est trop souvent oubliée !\n",
    "\n",
    "Voici votre meilleur allié : https://www.google.fr/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55cd050a-e887-4d14-a12e-6e01f72b7db7",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Une semaine de formation ne peut vous enseigner toutes les commandes et méthodes de la bioinformatique. Mais en comprendre la logique et acquérir les bases pour reformuler une question biologique en version informatique seront essentiels."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d335088a-8062-428b-845b-27593100529c",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": []
   },
   "source": [
    "Quand vos tuteurs ne seront plus là, voici également quelques exemples de _cheat sheet_ listant les commandes les plus courantes du <code>bash</code>. \n",
    "\n",
    "- [Un résumé des essentiels](https://www.educative.io/blog/bash-shell-command-cheat-sheet) \n",
    "- [Celle à imprimer et mettre dans son bureau](https://docslib.org/doc/8949368/beta-cheat-sheet-guide-to-general-linux-bash-and-slurm-commands-head-file-print-first-10-lines-of-file)\n",
    "- Une liste [très exhaustive](https://devhints.io/bash) et une liste [vraiment très très exhaustive](https://github.com/RehanSaeed/Bash-Cheat-Sheet)\n",
    "- [Plein d'exemples d'utilisation](https://www.laulem.com/en/dev/bash-memo.html)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
